[window] Prefer GTK icon theme over NET_WM_ICON

Kafva opened the pull request:
I have noticed that the `window` mode does not pick up certain icons from the icon theme set in `gtk-3.0.ini`. Looking at [window.c](https://github.com/davatorium/rofi/blob/fce721a1053ad6d861834e9f3a9c96c20bf1eb6d/source/modes/window.c#L1074) I can see that this is to be expected since the class name that is used to look up the icon name is not guaranteed to be correct.

However, currently the NETWM icon seems to override any GTK icons that are found, this patch is meant to address this.

Before/after patch using:
```ini
[Settings]
gtk-icon-theme-name=Papirus-Dark

# rofi -show window
```
([theme](https://gist.github.com/c7f9f31579f1e6b130a350fe05b6cec3))
![before](https://user-images.githubusercontent.com/36083692/189474505-a3e6caca-a57b-4b86-8423-94cba98fbc4a.png)
![after](https://user-images.githubusercontent.com/36083692/189474508-14317866-beb9-40b7-a8e0-6bb3724677f0.png)

Commit:
[window] Prefer GTK icon theme over NET_WM_ICON

DaveDavenport commented:
I don't think this patch implements correct behavior.
The icon the application explicitly sets for itself should get priority.
(some) Applications use this to relay extra information to the user.

Kafva commented:
Oh ye, I assume you mean like an email app setting "X unread messages" in the corner of the icon?
I would personally prefer that the icons always try to follow the theme but I guess that behaviour is worth preserving for other use cases.

DaveDavenport commented:
maybe get the applications to use the correct icon theme instead?

Kafva commented:
Hmmm ye, I guess the source of the problem isn't really rofi but how my environment sets (or does not set) the application icons that are sent. Definitely seems better to fix that instead if that is the case.

Kafva commented:
I tried my patch in a different window manager (awesomewm instead of bspwm) but saw the same behaviour as shown in my prior screenshots.

Were you implying earlier that each individual application (kitty, obs etc.) would need to be patched to provide an icon from the active GTK theme (in queries for their `NET_WM_ICON`) or is this something that the WM/DE has control over/should handle?

DaveDavenport commented:
I have no idea what a window manager has to do with what icon an application set for itself.

I don't know how applications exactly do this, I guess it depends on the application.

But I am not going to change rofi default behavior, that seems correct, because a user finds it 'ugly'.

Kafva commented:
Ye, the window manager is probably unrelated 😅

My main gripe with the current behaviour is that the `drun` mode _does_ show icons from the GTK theme, so it feels inconsistent that another mode shows different icons. 

But I am fine with using my fork instead if this is not desirable as the default.

DaveDavenport commented:
I don't think you read my reply right. I am not going to change the default behaviour, this does not rule out an config option to change the default behaviour.

Kafva commented:
Oh ok, I guess I will reopen the issue then.

Commit:
[window] Add -window-prefer-gtk-icons option

DaveDavenport commented:
I think gtk icons is not the right name,  `-window-prefer-icon-theme`  is better.

Commit:
[window] Rename option to -window-prefer-icon-theme

DaveDavenport commented on doc/rofi.1.markdown:
```diff
@@ -585,6 +585,9 @@ Default: *"wmctrl -i -R {window}"*
 
+`-window-prefer-icon-theme`
+
+Use icons from the current GTK theme in window mode.
```
It is not 'gtk theme'.  It is the icon theme set, this is not directly dependent on gtk. 
Rofi f.e. has it own set icon-theme option, and this will be used when set.

Kafva replied:
I changed it to  'icon theme', does it need to be more verbose in some other way?

DaveDavenport commented on source/modes/window.c:
```diff
@@ -1062,21 +1063,24 @@ static cairo_surface_t *_get_icon(const Mode *sw, unsigned int selected_line,
+  } else if (c->icon == NULL && c->icon_checked == FALSE) {
+    c->icon = get_net_wm_icon(rmpd->ids->array[selected_line], size);
+    c->icon_checked = TRUE;
+  }
```
Is this logic correct?  when `config.window_prefer_icon_theme` is not set, it will never fall back to the 'class icon' when the application has no `net_wm_icon` set.
It now always requests this icon, even when not used. This seems wasteful.

Kafva replied:
Tried to address this in [4bd7d3b](https://github.com/davatorium/rofi/pull/1702/commits/4bd7d3b59d70db4bc8e2f1f01305828c79f45e86)

Commit:
[docs] Change description of -window-prefer-icon-theme

Commit:
[window] Fix logic for -window-prefer-icon-theme

Only fetch the NET_WM_ICON and theme icon when their counterpart is NULL as a
fallback for each state of the `-window-prefer-icon-theme` flag.

DaveDavenport commented:
Going over the patch, there is a small possible glitching issue. It might show the application set by the application, then the one from the theme.
To fix this an extension needs to be made to rofi_icon_fetcher. I will look at this.

DaveDavenport commented:
Pushed the extension to rofi_icon_fetcher and directly implemented this request (this exposed another small issue, so decided to commit the whole).

Can you test?

DaveDavenport commented:
You can prefer the icon theme above the window set icon with the 'prefer-icon-theme' setting:

              configuration {
                window {
                    prefer-icon-theme: true;
                }
              }

       or pass -window-prefer-icon-theme true on command line.

Kafva commented:
Ye, tested the new option via CLI and from the config file, both seem to work as expected 🙂
